File: lib/shire/workspace.ex
defmodule Shire.Workspace do
  @moduledoc """
  Centralizes workspace path construction for all backends.
  Delegates to the configured VM implementation for the workspace root.
  """

  def root(project_id), do: vm().workspace_root(project_id)
  def agents_dir(project_id), do: Path.join(root(project_id), "agents")
  def agent_dir(project_id, agent_id), do: Path.join(agents_dir(project_id), "#{agent_id}")
  def shared_dir(project_id), do: Path.join(root(project_id), "shared")
  def env_path(project_id), do: Path.join(root(project_id), ".env")
  def scripts_dir(project_id), do: Path.join(root(project_id), ".scripts")
  def script_path(project_id, name), do: Path.join(scripts_dir(project_id), name)
  def runner_dir(project_id), do: Path.join(root(project_id), ".runner")
  def peers_path(project_id), do: Path.join(root(project_id), "peers.yaml")
  def project_doc_path(project_id), do: Path.join(root(project_id), "PROJECT.md")

  def attachments_dir(project_id, agent_id),
    do: Path.join(agent_dir(project_id, agent_id), "attachments")

  def attachment_dir(project_id, agent_id, attachment_id),
    do: Path.join(attachments_dir(project_id, agent_id), attachment_id)

  def attachment_path(project_id, agent_id, attachment_id, filename),
    do: Path.join(attachment_dir(project_id, agent_id, attachment_id), filename)

  defp vm, do: Application.get_env(:shire, :vm, Shire.VirtualMachineSprite)
end

File: src/services/workspace.ts
import { join } from "path";
import { homedir } from "os";
import { mkdirSync } from "fs";

export function projectsDir(): string {
  return process.env.SHIRE_PROJECTS_DIR || join(homedir(), ".shire", "projects");
}

export function root(projectId: string): string {
  return join(projectsDir(), projectId);
}

export function agentsDir(projectId: string): string {
  return join(root(projectId), "agents");
}

export function agentDir(projectId: string, agentId: string): string {
  return join(agentsDir(projectId), agentId);
}

export function inboxDir(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "inbox");
}

export function outboxDir(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "outbox");
}

export function scriptsDir(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "scripts");
}

export function documentsDir(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "documents");
}

export function attachmentsDir(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "attachments");
}

export function attachmentDir(projectId: string, agentId: string, attachmentId: string): string {
  return join(attachmentsDir(projectId, agentId), attachmentId);
}

export function attachmentPath(
  projectId: string,
  agentId: string,
  attachmentId: string,
  filename: string,
): string {
  return join(attachmentDir(projectId, agentId, attachmentId), filename);
}

export function sharedDir(projectId: string): string {
  return join(root(projectId), "shared");
}

export function runnerDir(projectId: string): string {
  return join(root(projectId), ".runner");
}

export function peersPath(projectId: string): string {
  return join(root(projectId), "peers.yaml");
}

export function projectDocPath(projectId: string): string {
  return join(root(projectId), "PROJECT.md");
}

export function recipePath(projectId: string, agentId: string): string {
  return join(agentDir(projectId, agentId), "recipe.yaml");
}

export function ensureProjectDirs(projectId: string): void {
  mkdirSync(root(projectId), { recursive: true });
  mkdirSync(sharedDir(projectId), { recursive: true });
  mkdirSync(runnerDir(projectId), { recursive: true });
}

export function ensureAgentDirs(projectId: string, agentId: string): void {
  mkdirSync(agentDir(projectId, agentId), { recursive: true });
  mkdirSync(inboxDir(projectId, agentId), { recursive: true });
  mkdirSync(outboxDir(projectId, agentId), { recursive: true });
  mkdirSync(scriptsDir(projectId, agentId), { recursive: true });
  mkdirSync(documentsDir(projectId, agentId), { recursive: true });
  mkdirSync(attachmentsDir(projectId, agentId), { recursive: true });
  mkdirSync(join(attachmentsDir(projectId, agentId), "outbox"), {
    recursive: true,
  });
}
